File: parsers/dart/lib/src/meta.dart
// SYNX metadata + parse-result types. Mirrors Rust value.rs.

import 'value.dart';

/// Constraints from `[min:3, max:30, required, type:int, pattern:..., enum:a|b, readonly]`.
class SynxConstraints {
  double? min;
  double? max;
  String? typeName;
  bool required = false;
  bool readonly = false;
  String? pattern;
  List<String>? enumValues;

  SynxConstraints();

  bool get hasAny =>
      min != null ||
      max != null ||
      typeName != null ||
      required ||
      readonly ||
      pattern != null ||
      enumValues != null;
}

/// Marker / args / type-hint / constraints bundle attached to one active-mode field.
class SynxMeta {
  List<String> markers = [];

  /// One arg per marker (same length as `markers`).
  List<String> args = [];
  String? typeHint;
  SynxConstraints? constraints;

  SynxMeta();

  bool hasMarker(String name) => markers.contains(name);

  /// Index of marker `name` in the chain, or -1.
  int markerIndex(String name) {
    for (var i = 0; i < markers.length; i++) {
      if (markers[i] == name) return i;
    }
    return -1;
  }
}

typedef SynxMetaMap = Map<String, SynxMeta>;
typedef SynxMetadataTree = Map<String, SynxMetaMap>;

class SynxIncludeDirective {
  final String path;
  final String alias;
  const SynxIncludeDirective(this.path, this.alias);
}

class SynxUseDirective {
  final String pkg;
  final String alias;
  const SynxUseDirective(this.pkg, this.alias);
}

class SynxParseResult {
  SynxValue root;
  SynxMode mode;
  bool locked;
  bool tool;
  bool schema;
  bool llm;
  SynxMetadataTree metadata;
  List<SynxIncludeDirective> includes;
  List<SynxUseDirective> uses;

  SynxParseResult({
    SynxValue? root,
    this.mode = SynxMode.static_,
    this.locked = false,
    this.tool = false,
    this.schema = false,
    this.llm = false,
    SynxMetadataTree? metadata,
    List<SynxIncludeDirective>? includes,
    List<SynxUseDirective>? uses,
  })  : root = root ?? synxObject(),
        metadata = metadata ?? {},
        includes = includes ?? [],
        uses = uses ?? [];
}

File: parsers/go/meta.go
package synx

// Mode is the file-level parser mode.
type Mode uint8

const (
	ModeStatic Mode = iota
	ModeActive
)

// Constraints from `[min:3, max:30, required, type:int, pattern:..., enum:a|b, readonly]`.
type Constraints struct {
	Min        *float64
	Max        *float64
	TypeName   string // empty when absent
	Required   bool
	Readonly   bool
	Pattern    string // empty when absent
	EnumValues []string
}

func (c Constraints) HasAny() bool {
	return c.Min != nil || c.Max != nil || c.TypeName != "" ||
		c.Required || c.Readonly || c.Pattern != "" || c.EnumValues != nil
}

// Meta is the marker / args / type-hint / constraint bundle attached to one
// active-mode field.
type Meta struct {
	Markers     []string
	Args        []string
	TypeHint    string // empty when absent
	Constraints *Constraints
}

func (m Meta) HasMarker(name string) bool {
	for _, x := range m.Markers {
		if x == name {
			return true
		}
	}
	return false
}

// MarkerIndex returns the index of `name` in the chain, or -1.
func (m Meta) MarkerIndex(name string) int {
	for i, x := range m.Markers {
		if x == name {
			return i
		}
	}
	return -1
}

// MetaMap is `key -> Meta` for one object level.
type MetaMap map[string]Meta

// MetadataTree is keyed by dot-path prefix ("" for the root level).
type MetadataTree map[string]MetaMap

// IncludeDirective parsed from `!include path [alias]`.
type IncludeDirective struct {
	Path  string
	Alias string
}

// UseDirective parsed from `!use @scope/name [as alias]`.
type UseDirective struct {
	Package string
	Alias   string
}

// ParseResult is the complete output of Parse — root tree, mode flags,
// metadata, include/use directives.
type ParseResult struct {
	Root     Value
	Mode     Mode
	Locked   bool
	Tool     bool
	Schema   bool
	Llm      bool
	Metadata MetadataTree
	Includes []IncludeDirective
	Uses     []UseDirective
}
